lib/fswatcher/fswatcher.go

```diff
@@ -0,0 +1,577 @@
+	for _, childDir := range dir.dirs {
+		for path, event := range w.popOldEvents(childDir, currTime) {
+			oldEvents[path] = event
+		}
+	}
+	for path, event := range dir.events {
+		if w.isOld(event, currTime) {
+			oldEvents[path] = event
+			delete(dir.events, path)
+		}
+	}
+	if dir.parentDir != nil && dir.childCount() == 0 {
+		dir.parentDir.removeEmptyDir(dir.path)
+	}
+	return oldEvents
+}
+
+func (w *watcher) isOld(ev *event, currTime time.Time) bool {
+	// Deletes should always be scanned last, therefore they are always
+	// delayed by letting them time out (see below).
+	// An event that has not registered any new modifications recently is scanned.
+	// w.notifyDelay is the user facing value signifying the normal delay between
+	// a picking up a modification and scanning it. As scheduling scans happens at
+	// regular intervals of w.notifyDelay the delay of a single event is not exactly
+	// w.notifyDelay, but lies in in the range of 0.5 to 1.5 times w.notifyDelay.
+	if ev.evType == nonRemove && 2*currTime.Sub(ev.lastModTime) > w.notifyDelay {
+		return true
+	}
+	// When an event registers repeat modifications or involves removals it
+	// is delayed to reduce resource usage, but after a certain time (notifyTimeout)
+	// passed it is scanned anyway.
+	return currTime.Sub(ev.firstModTime) > w.notifyTimeout
+}
+
+func (w *watcher) updateInProgressSet(event events.Event) {
+	if event.Type == events.ItemStarted {
+		path := event.Data.(map[string]string)["item"]
+		w.inProgress[path] = struct{}{}
+	} else if event.Type == events.ItemFinished {
+		path := event.Data.(map[string]interface{})["item"].(string)
+		delete(w.inProgress, path)
+	}
+}
+
+func (w *watcher) pathInProgress(path string) bool {
+	_, exists := w.inProgress[path]
+	return exists
+}
+
+func (w *watcher) UpdateIgnores(ignores *ignore.Matcher) {
+	l.Debugln(w, "Ignore patterns update")
+	select {
+	case w.folderIgnoresUpdate <- ignores:
+	case <-w.stop:
+	}
+}
+
+func (w *watcher) String() string {
+	return fmt.Sprintf("fswatcher/%s:", w.folderCfg.Description())
+}
+
+func (w *watcher) eventType(notifyType notify.Event) eventType {
+	if notifyType&w.removeEventMask() != 0 {
+		return remove
+	}
+	return nonRemove
+}
+
+func (w *watcher) VerifyConfiguration(from, to config.Configuration) error {
+	return nil
+}
+
+func (w *watcher) CommitConfiguration(from, to config.Configuration) bool {
+	for _, folderCfg := range to.Folders {
+		if folderCfg.ID == w.folderID {
+			select {
+			case w.configUpdate <- folderCfg:
+			case <-w.stop:
+			}
+			return true
+		}
+	}
+	// Nothing to do, model will soon stop this service
+	return true
+}
+
+func (w *watcher) updateConfig(folderCfg config.FolderConfiguration) {
+	w.folderPath = filepath.Clean(folderCfg.Path())
+	w.notifyDelay = time.Duration(folderCfg.FSWatcherDelayS) * time.Second
+	w.notifyTimeout = notifyTimeout(folderCfg.FSWatcherDelayS)
+	w.folderCfg = folderCfg
+}
+
+func (dir *eventDir) eventCount() int {
```

Comment:
I placed all struct definitions before they are first used and ordered functions by their structs - starting with those on `*watcher`. Should I move `*eventDir` functions to the beginning or the struct definition back?